### PR TITLE
only release the gurobi license if all instances have been deleted

### PR DESCRIPTION
**Pull request by michaelbynum**
## Changes proposed in this PR:
- only release the gurobi license if all instances have been deleted

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/appsi/solvers/gurobi.py [hunk 1]
```diff
@@ -274,8 +276,10 @@ def release_license(self):
                 gurobipy.disposeDefaultEnv()
 
     def __del__(self):
-        if not python_is_shutting_down():
-            self.release_license()
+        self._num_instances -= 1
+        if self._num_instances == 0:
+            if not python_is_shutting_down():
```
Comment (jsiirola) [inline pyomo/contrib/appsi/solvers/gurobi.py:281]: You almost certainly need to check python_is_shutting_down before _num_instances (an when python is shutting down, the _num_instances attribute may have already been released)

Comment (michaelbynum) [reply pyomo/contrib/appsi/solvers/gurobi.py]: Great point. Thanks!

**Comment by jsiirola** [comment]
I am a little confused by this.  Doesn't each APPSI solver interface consume a single license?  Why do we only release the license for the last APPSI instance to be disposed?

**Comment by michaelbynum** [comment]
No, you can create as many Gurobi models as you want in a single process with a single license.

**Comment by jsiirola** [comment]
Got it, Thanks!.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**